pkg/util/validation/limits.go

```diff
@@ -885,3 +880,108 @@ func EnabledByAnyTenant(tenantIDs []string, f func(string) bool) bool {
 	}
 	return false
 }
+
+// MustRegisterExtension registers the extensions type with given name
+// and returns a function to get a pointer to the extensions E from a *Limits instance.
+// The name will be used as YAML/JSON key to decode the extensions.
+// This method is not thread safe and should be called only during package initialization.
+// Registering same name twice will cause a panic.
+func MustRegisterExtension[E any](name string) func(*Limits) *E {
+	if name == "" {
+		panic("extension name cannot be empty")
+	}
+	if _, ok := standardLimitsYAMLJSONKeys[name]; ok {
+		panic(fmt.Errorf("extension %s cannot be registered because it's a standard limits field", name))
+	}
+	if _, ok := registeredExtensionsIndexes[name]; ok {
+		panic(fmt.Errorf("extension %s already registered", name))
+	}
+	registeredExtensionsIndexes[name] = len(registeredExtensionsIndexes)
+
+	limitsExtensionsFields = append(limitsExtensionsFields, reflect.StructField{
+		Name: strings.ToUpper(name),
+		Type: reflect.TypeOf(new(E)),
+		Tag:  reflect.StructTag(fmt.Sprintf(`yaml:"%s" json:"%s"`, name, name)),
+	})
+
+	return func(l *Limits) *E {
+		if e, ok := l.extensions[name]; ok {
+			return e.(*E)
+		}
+		return nil
+	}
+}
+
+var standardLimitsYAMLJSONKeys = map[string]struct{}{}
+
+func init() {
+	limitsType := reflect.TypeOf(Limits{})
+	for i := 0; i < limitsType.NumField(); i++ {
+		// yamlKey/jsonKey could be empty, but we also shouldn't allow registering a field with an empty name, so just add it to the map.
+		yamlKey, _, _ := strings.Cut(limitsType.Field(i).Tag.Get("yaml"), ",")
+		josnKey, _, _ := strings.Cut(limitsType.Field(i).Tag.Get("json"), ",")
```

Comment:
```suggestion
		jsonKey, _, _ := strings.Cut(limitsType.Field(i).Tag.Get("json"), ",")
```

Reply:
Oh dear... Fixed both names in https://github.com/grafana/mimir/pull/4397/commits/feb20eee460fd596c62fc80d0d1bb1613245b91c